nginx/nginx.go

```diff
@@ -349,60 +350,99 @@ func (lb *nginxLoadBalancer) createConfig(update controller.IngressUpdate) ([]by
+			ID:     upstreamID(ingressEntry),
 			Server: fmt.Sprintf("%s:%d", ingressEntry.ServiceAddress, ingressEntry.ServicePort),
 		}
+		idToUpstream[upstream.ID] = upstream
+	}
+
+	var sortedUpstreams []*upstream
+	for _, upstream := range idToUpstream {
+		sortedUpstreams = append(sortedUpstreams, upstream)
+	}
+
+	sort.Sort(upstreams(sortedUpstreams))
+	return sortedUpstreams
+}
+
+func upstreamID(e controller.IngressEntry) string {
+	return fmt.Sprintf("%s.%s.%d", e.Namespace, e.ServiceAddress, e.ServicePort)
+}
+
+type servers []*server
+
+func (s servers) Len() int           { return len(s) }
+func (s servers) Less(i, j int) bool { return s[i].ServerName < s[j].ServerName }
+func (s servers) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+type locations []location
+
+func (l locations) Len() int           { return len(l) }
+func (l locations) Less(i, j int) bool { return l[i].Path < l[j].Path }
+func (l locations) Swap(i, j int)      { l[i], l[j] = l[j], l[i] }
+
+type pathSet map[string]bool
+
+func createServerEntries(update controller.IngressUpdate) []*server {
+	hostToNginxEntry := make(map[string]*server)
+	hostToPaths := make(map[string]pathSet)
+
+	for _, ingressEntry := range update.Entries {
+		serverEntry, exists := hostToNginxEntry[ingressEntry.Host]
+		if !exists {
+			serverEntry = &server{ServerName: ingressEntry.Host}
+			hostToNginxEntry[ingressEntry.Host] = serverEntry
+			hostToPaths[ingressEntry.Host] = make(map[string]bool)
+		}
+
+		nginxPath := createNginxPath(ingressEntry.Path)
 		location := location{
 			Path:                    nginxPath,
-			UpstreamID:              upstream.ID,
+			UpstreamID:              upstreamID(ingressEntry),
 			Allow:                   ingressEntry.Allow,
 			StripPath:               ingressEntry.StripPaths,
 			BackendKeepaliveSeconds: ingressEntry.BackendKeepAliveSeconds,
 		}
 
-		entry, exists := hostToNginxEntry[ingressEntry.Host]
-		if !exists {
-			entry = &nginxEntry{ServerName: ingressEntry.Host}
-			hostToNginxEntry[ingressEntry.Host] = entry
-			nginxEntries = append(nginxEntries, entry)
-			hostToPaths[ingressEntry.Host] = make(map[string]struct{})
-		}
-
 		paths := hostToPaths[ingressEntry.Host]
-		if _, exists := paths[location.Path]; exists {
-			log.Infof("Ignoring '%s' because it duplicates the host/path of a previous entry", ingressEntry.Name)
+		if paths[location.Path] {
+			log.Infof("Ignoring '%s' because it duplicates the host/path of a previous entry", ingressEntry.NamespaceName())
 			continue
 		}
-		paths[location.Path] = struct{}{}
+		paths[location.Path] = true
+
+		serverEntry.Name += " " + ingressEntry.NamespaceName()
+		serverEntry.Locations = append(serverEntry.Locations, location)
+	}
 
-		entry.Name += " " + ingressEntry.Name
-		entry.Upstreams = append(entry.Upstreams, upstream)
-		entry.Locations = append(entry.Locations, location)
-		upstreamIndex++
+	var serverEntries []*server
+	for _, serverEntry := range hostToNginxEntry {
```

Comment:
is the extra sorting to avoid an update when nothing has changed and go gives it back in a different order?